src/main/java/org/terracotta/tinypounder/TinyPounderMainUI.java

```diff
@@ -151,33 +154,47 @@ protected void init(VaadinRequest vaadinRequest) {
         2, 2, TimeUnit.SECONDS);
   }
 
+  private void addExitCloseTab() {
+    VerticalLayout exitLayout = new VerticalLayout();
+    exitLayout.addComponentsAndExpand(new Label("We hope you had fun using the TinyPounder, it is now shutdown, " +
+        "as well as all the DatasetManagers, CacheManagers, and Terracotta servers you started with it"));
+    TabSheet.Tab tab = mainLayout.addTab(exitLayout, "EXIT : Close TinyPounder");
+    tab.setStyleName("tab-absolute-right");
+    mainLayout.addSelectedTabChangeListener(tabEvent -> {
+      if (tabEvent.getTabSheet().getSelectedTab().equals(tab.getComponent())) {
+        new Thread(() -> {
+          runningServers.values().forEach(RunningServer::stop);
+          consoleRefresher.cancel(true);
+          SpringApplication.exit(appContext);
+        }).start();
+      }
+    });
+  }
+
   private void updateKitControls() {
     if (kitAwareClassLoaderDelegator.isEEKit()) {
       if (kitPathLayout.getRows() == 1) {
-        TextField path = new TextField();
-        path.setWidth(100, Unit.PERCENTAGE);
-        path.setValue(settings.getLicensePath() == null ? "" : settings.getLicensePath());
-        path.addValueChangeListener(event -> settings.setLicensePath(event.getValue()));
-        path.setPlaceholder("License location");
-        Upload upload = new Upload();
-        upload.setReceiver((Upload.Receiver) (filename, mimeType) -> {
-          File tmpDir = new File(System.getProperty("java.io.tmpdir"));
-          temporaryUploadedLicenseFile = new File(tmpDir, filename);
-          temporaryUploadedLicenseFile.deleteOnExit();
+        final TextField licensePath = new TextField();
+        licensePath.setWidth(100, Unit.PERCENTAGE);
+        licensePath.setValue(settings.getLicensePath() == null ? "" : settings.getLicensePath());
+        licensePath.addValueChangeListener(event -> settings.setLicensePath(event.getValue()));
+        licensePath.setPlaceholder("License location");
+        licensePath.addValueChangeListener(event -> {
```

Comment:
there are 2 `addValueChangeListener` ;-) It works ?

Reply:
yes it works !
actually I've used it somewhere else in this PR, where I clearly wanted the 2 lambdas to be separate.
in that case, that does not make sense, let me repush